### DISCUSSION TITLE
How to inherit secrets when reusing workflows in Github Actions

### DISCUSSION THREAD

**Discussion by iagobruno** [community/community#23107] [New to GitHub]
<p>It was recently <a href="https://github.blog/changelog/2022-05-03-github-actions-simplify-using-secrets-with-reusable-workflows/" rel="noopener nofollow ugc">posted on the Github changelog blog</a> that it’s now possible to inherit secrets between workflows and I went to understand better how it works in the link at the end of the article (<a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow" class="inline-onebox" rel="noopener nofollow ugc">Reusing workflows - GitHub Docs</a>).</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/2/624ebbc33b9bbc18d73249b90fdf10540e30d552.png" data-download-href="/uploads/short-url/e1FtCqjoU0jfdlUiqK0gnEVOw0i.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/3616259/181074879-836f618c-5a4b-4664-a3a4-8216a5d7519b.png" alt="image" data-base62-sha1="e1FtCqjoU0jfdlUiqK0gnEVOw0i" width="690" height="398" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/2/624ebbc33b9bbc18d73249b90fdf10540e30d552_2_690x398.png, https://aws1.discourse-cdn.com/github/optimized/3X/6/2/624ebbc33b9bbc18d73249b90fdf10540e30d552_2_1035x597.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/6/2/624ebbc33b9bbc18d73249b90fdf10540e30d552.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/2/624ebbc33b9bbc18d73249b90fdf10540e30d552_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1136×656 48.2 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>I tried to use it in a project that I’m working on in the same way as it’s explained in the documentation but it didn’t work. See the print below.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/6/8/68af0b873819df932d8de3b404e55c1689331f5d.png" data-download-href="/uploads/short-url/eW4GJcsASvwn2NzobXH9Jg3jNCd.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/3616259/181074880-d38fa8f8-748a-4b70-ab4f-e7df1a356c47.png" alt="image" data-base62-sha1="eW4GJcsASvwn2NzobXH9Jg3jNCd" width="690" height="457" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/6/8/68af0b873819df932d8de3b404e55c1689331f5d_2_690x457.png, https://aws1.discourse-cdn.com/github/optimized/3X/6/8/68af0b873819df932d8de3b404e55c1689331f5d_2_1035x685.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/6/8/68af0b873819df932d8de3b404e55c1689331f5d.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/6/8/68af0b873819df932d8de3b404e55c1689331f5d_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1271×842 26.8 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>I also tried using <code>secrets: inherit</code> in the <code>on.workflow_call</code> property in the ci.yml file but that didn’t work either.</p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/7/8/7877865d2cb5937542d02580b46f8f8d80893d9c.png" data-download-href="/uploads/short-url/hbHnohhAVoxeb2eE6EuHsIpGMj2.png?dl=1" title="image" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/3616259/181074881-87e580ca-6a9f-45ec-94c2-f3f1772281de.png" alt="image" data-base62-sha1="hbHnohhAVoxeb2eE6EuHsIpGMj2" width="690" height="347" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/7/8/7877865d2cb5937542d02580b46f8f8d80893d9c_2_690x347.png, https://aws1.discourse-cdn.com/github/optimized/3X/7/8/7877865d2cb5937542d02580b46f8f8d80893d9c_2_1035x520.png 1.5x, https://aws1.discourse-cdn.com/github/original/3X/7/8/7877865d2cb5937542d02580b46f8f8d80893d9c.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/7/8/7877865d2cb5937542d02580b46f8f8d80893d9c_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use href="#far-image"></use></svg><span class="filename">image</span><span class="informations">1310×659 21.4 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use href="#discourse-expand"></use></svg>
</div></a></div>
<p>Can someone explain why it’s not working or what i’m doing wrong?<br>
Thanks</p>

**Comment by airtower-luna** [comment, marked as answer]
<p>The first one doesn’t work because you’re trying to call the reusable workflow in a <em>step</em>, which isn’t possible (that’s for custom actions). A reusable workflow must be called at the job level.</p>
<p>Point three under <a href="https://docs.github.com/en/actions/using-workflows/reusing-workflows#using-inputs-and-secrets-in-a-reusable-workflow">Using inputs and secrets in a reusable workflow</a> has an example. I notice it also says there:</p>
<blockquote>
<p>Workflows that call reusable workflows in the same organization or enterprise can use the <code>inherit</code> keyword to implicitly pass the secrets.</p>
</blockquote>
<p>I hope that doesn’t mean you can’t use that for workflows owned by the same personal account. :thinking:</p>

**Comment by jsoref** [comment]
<blockquote>
<p>I hope that doesn’t mean you can’t use that for workflows owned by the same personal account. :thinking:</p>
</blockquote>
<p>It almost certainly means exactly that.</p>
<p>Just create an organization. They’re free. I have a couple.</p>

**Comment by iagobruno** [comment]
<p>Thanks, now it’s working as expected. It was just a lack of attention.</p>
<p>And about the concern of the repository being in an organization or a personal account, I didn’t have any issues with that.</p>
<p>The final file version for future visitors:</p>

```yml
on:
  push:
    branches: [main]

jobs:
  tests:
    uses: ./.github/workflows/ci.yml
    secrets: inherit

  deploy:
    runs-on: ubuntu-latest
    needs: tests
    steps:
      - uses: actions/checkout@v3
      - name: Deploy
        run: echo 'DEPLOY NEW VERSION!'
```

**Reply by ssbarnea** [reply]
You should better paste valid yaml for "future visitors" ;)

**Comment by SgtPooki** [comment]
<p>What does your reusable workflow look like? inheriting secrets seems so broken… Errors suggest I have to specify exact secrets, but documentation says I can inherit. <a href="https://github.com/ipfs-shipyard/pinning-service-compliance/pull/128" class="inline-onebox" rel="noopener nofollow ugc">feat: compose github actions by SgtPooki · Pull Request #128 · ipfs-shipyard/pinning-service-compliance · GitHub</a></p>

**Comment by ffisc** [comment]
Hopping onto the thread, a question about using `inherit` in parallel with explicitly passing secrets.
Suppose repo `a` has a reusable workflow `RWA` which requires secrets `s1`, `s2` and `s3`
In repo `b`, I'd like to call `RWA`. However, while repo `b` has `s1`, `s2` stored as such, in this repo `s3` is stored as `s_3`.
What I'd like to do is use `inherit` (so I don't have to pass `s1` and `s2` explicitly) *and* specify `s3: ${{ secrets.s_3 }}`.
Something like:
```
jobs:
  tests:
    uses: Org/a/.github/workflows/RWA.yml
    secrets:
      inherit  # inherit s1 and s2
      s3: ${{ secrets.s_3 }}  # explicitly assign s3
```
It doesn't seem possible to achieve this, right? Is there any workaround?

**Reply by jsoref** [reply]
That docs say `alternatively`, so I'd expect that it's an exclusive or.
https://docs.github.com/en/actions/using-workflows/reusing-workflows#passing-secrets-to-nested-workflows

Given how I'd imagine implementing this, I wouldn't expect what you're looking for to be supported. So the workaround is to just be explicit (which is probably better for people trying to read your workflows later anyway).

**Reply by ffisc** [reply]
Thanks. I think in the case where there are 3 secrets then being explicit doesn't hurt, but if there are 10 and only 1 has a different name then one might start scratching his head... but yes this might be en edge case.

**Reply by jsoref** [reply]
As someone who regularly has to pick up after other people have left or deal w/ spelling inconsistencies, it's really helpful if people tie everything together by showing their dependencies. I much prefer that to the case where there's something magical going on where I change something in one place and it isn't remotely obvious that it's being used somewhere else...

**Reply by tibistibi** [reply]
i have to a lot of secrets which just pass to the used workflow with just two which change (depending on the environment).

this would work perectly:

jobs:
  tests:
    uses: Org/a/.github/workflows/RWA.yml
    secrets:
      inherit: true  # inherit s1 and s2
      s3: ${{ secrets.s_3 }}  # explicitly assign s3

**Reply by nvx** [reply]
> i have to a lot of secrets which just pass to the used workflow with just two which change (depending on the environment).
> 
> this would work perectly:
> 
> jobs: tests: uses: Org/a/.github/workflows/RWA.yml secrets: inherit: true # inherit s1 and s2 s3: ${{ secrets.s_3 }} # explicitly assign s3

I've tried this but it errors out saying inherit isn't defined in the reusable workflow, if I add it then it errors out that it's a bool value instead of a string. Does thist still work?

**Reply by Borda** [reply]
It used to work, but recently I've seen crashes on our jobs too

**Comment by iamintown** [comment]
caller-workflow -> reusable-workflow-> composite-action.yml 
my composite workflow from reusable is not get identity custom-actions/s3-create
Run set -e
✅ Using AWS region: us-east-1

Unable to locate credentials. You can configure credentials by running "aws configure".
⚠️ Unable to verify caller identity
🪣 Processing bucket: tec-dev-01
🛠️ Creating bucket: tec-dev-ut-01

Unable to locate credentials. You can configure credentials by running "aws configure".
Error: Process completed with exit code 253.
